### PR TITLE
Added initial support for decision trees.

### PR DESCRIPTION
**Pull request by mossbanay**
Re my comment on #7. This seems to be able to fit trees correctly but is much slower than the sklearn implementation.

The main algorithm for fitting to a masked subset of the data is in TreeNode::fit. It iterates over the features in the dataset and creates a `SplitIterator` over the possible options to split values on (this means we only have to sort along each feature once at each level). It then computes the impurity criterion for each of the possible splits along that feature. After the split that reduces the impurity the most is found, masks are created for the nodes that belong to the left or right children.

Overall this should be O(k n log(n)) at each level which matches the complexity stated in the [sklearn docs](https://scikit-learn.org/stable/modules/tree.html#complexity). My guess is a *big* speedup could come from someone with more experience with ndarray looking over some of the code since I'm not familiar with it.

There's an included benchmark using criterion. Just run `cargo bench decision_tree` in the linfa-trees folder.

### CODE REVIEW

**Comment by LukeMathWalker** [comment]
Amazing :grin: I'll have a look over the weekend to see if something strikes me as a clear bottleneck :+1:

**Comment by leaen** [comment]
I've just made a change to how the masks on the sub-trees are represented. Previously I passed a `Vec<usize>` containing the indices of the elements of the sub-tree around, but this required a sort over each feature axis at each level. I've swapped to using a `Vec<bool>` as a binary mask inside a `RowMask` struct. This way, I precompute where each row lands when sorting along a particular feature in a `SortedIndex` struct and since this never changes, you only need sort along each feature once.

It now takes ~500ms to fit a decision tree on ~400k samples on a i5-7200U, although since I haven't used the python bindings before I haven't been able to do a fair comparison with sklearn.

**Comment by bytesnake** [comment]
here some questions, because I never learned algorithms for decision trees:
 * for a feature bagging decision tree (like random forest) it can be just re-used and called with a initialized mask for the root tree right?
 * how much effort would it be to generalize this to regression? 
 * would it make sense to create a `metrics` module in the base crate for accuracy and all its relatives?

File: linfa-trees/examples/decision_tree.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
+use linfa_clustering::generate_blobs;
+use linfa_trees::{DecisionTree, DecisionTreeParams, SplitQuality};
+use ndarray::{array, Array, ArrayBase, Data, Ix1};
+use ndarray_rand::rand::SeedableRng;
+use rand_isaac::Isaac64Rng;
+use std::iter::FromIterator;
+
+fn accuracy(
+    labels: &ArrayBase<impl Data<Elem = u64> + Sync, Ix1>,
```
[the remark this answers is not in the document]
Comment (mossbanay) [reply linfa-trees/examples/decision_tree.rs:9]: Honestly I don't think it's required anywhere here. I must've copied it over when initially looking at how to use ndarray.

File: linfa-trees/src/decision_trees/algorithm.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+
+impl RowMask {
+    fn all(n_samples: u64) -> Self {
+        RowMask {
+            mask: vec![true; n_samples as usize],
+            n_samples: n_samples,
+        }
+    }
+
+    fn of_vec(mask: Vec<bool>) -> Self {
```
[the remark this answers is not in the document]
Comment (mossbanay) [reply linfa-trees/src/decision_trees/algorithm.rs:17]: I've moved it to the tests module

**Comment by mossbanay** [comment]
> here some questions, because I never learned algorithms for decision trees:
> 
>     * for a feature bagging decision tree (like random forest) it can be just re-used and called with a initialized mask for the root tree right?

Sort of yeah. For random forests we'd need to add a mask on which features of the dataset can be used (or pass in a dataset with a subset of the columns available). 

>     * how much effort would it be to generalize this to regression?

Very little, you'd just need to change `prediction_for_rows` to take the average of the y values in the rows from the mask instead of the modal class.

>     * would it make sense to create a `metrics` module in the base crate for accuracy and all its relatives?

Yep I'm strongly in favour of this since it'll also encourage standardisation of the predictions too.

**Comment by mossbanay** [comment]
Still going through your feedback, just had a very busy week and not much time to look over this :(

**Comment by bytesnake** [comment]
sure take your time, I'm always trying to spend 1/2 - 1 hour each day, but sometimes it's difficult

File: linfa-trees/src/decision_trees/algorithm.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            hyperparameters,
+            root_node,
+        }
+    }
+
+    /// Make predictions for each row of a matrix of features `x`.
+    pub fn predict(&self, x: &ArrayBase<impl Data<Elem = f64>, Ix2>) -> Array1<u64> {
+        let mut preds = vec![];
+
+        for row in x.genrows().into_iter() {
```
[the remark this answers is not in the document]
Comment (mossbanay) [reply linfa-trees/src/decision_trees/algorithm.rs:243]: Very nice, ty

**Comment by mossbanay** [comment]
Followed up the remaining feedback and merged in master

**Comment by bytesnake** [comment]
if you have interest in an extension for `RandomForest`, any PR is welcome
